magefile.go

```diff
@@ -0,0 +1,496 @@
+//go:build mage
+// +build mage
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/sh"
+	"go.uber.org/multierr"
+)
+
+var _depBin = filepath.Join(_dependencyDir, "bin")
+
+var _dependencyDir = func() string {
+	if dir, ok := os.LookupEnv("DEPENDENCY_DIR"); ok {
+		return dir
+	}
+
+	return filepath.Join(_projectRoot, ".cache", "dependencies")
+}()
+
+var _projectRoot = func() string {
+	if root, ok := os.LookupEnv("PROJECT_ROOT"); ok {
+		return root
+	}
+
+	root, err := sh.Output("git", "rev-parse", "--show-toplevel")
+	if err != nil {
+		panic("failed to get working directory")
+	}
+
+	return root
+}()
+
+var _tag = func() string {
```

Comment:
In this domain it is always a tag (at least to keep it consistent with the `Makefile`). If in the future it's not a tag it would be refactored to an independent function from which the top level names are populated.